src/julia_threads.h

```diff
@@ -0,0 +1,361 @@
+    // non GC-safe region.
+    if (old_state && !state)
+        jl_gc_safepoint();
+    return old_state;
+}
+#endif // ifndef JULIA_ENABLE_THREADING
+STATIC_INLINE int8_t jl_gc_state_save_and_set(int8_t state)
+{
+    return jl_gc_state_set(state, jl_gc_state());
+}
+#define jl_gc_unsafe_enter() jl_gc_state_save_and_set(0)
+#define jl_gc_unsafe_leave(state) ((void)jl_gc_state_set((state), 0))
+#define jl_gc_safe_enter() jl_gc_state_save_and_set(JL_GC_STATE_SAFE)
+#define jl_gc_safe_leave(state) ((void)jl_gc_state_set((state), JL_GC_STATE_SAFE))
+
+// Locks
+#ifdef JULIA_ENABLE_THREADING
+#define JL_DEFINE_MUTEX(m)                                                \
+    uint64_t volatile m ## _mutex = 0;                                    \
+    int32_t m ## _lock_count = 0;                                         \
+    void jl_unlock_## m ## _func(void)                                    \
+    {                                                                     \
+        JL_UNLOCK_RAW(m);                                                 \
+    }
+
+#define JL_DEFINE_MUTEX_EXT(m)                                            \
+    extern uint64_t volatile m ## _mutex;                                 \
+    extern int32_t m ## _lock_count;                                      \
+    void jl_unlock_## m ## _func(void);
+
+#define JL_LOCK_WAIT(m, wait_ex) do {                                   \
+        if (m ## _mutex == uv_thread_self()) {                          \
+            ++m ## _lock_count;                                         \
+        }                                                               \
+        else {                                                          \
+            for (;;) {                                                  \
+                if (m ## _mutex == 0 &&                                 \
+                    jl_atomic_compare_exchange(&m ## _mutex, 0,         \
+                                               uv_thread_self()) == 0) { \
+                    m ## _lock_count = 1;                               \
+                    break;                                              \
+                }                                                       \
+                wait_ex;                                                \
+                jl_cpu_pause();                                         \
+            }                                                           \
+        }                                                               \
+    } while (0)
+
+#define JL_UNLOCK_RAW(m) do {                                           \
+        if (m ## _mutex == uv_thread_self()) {                          \
+            --m ## _lock_count;                                         \
+            if (m ## _lock_count == 0) {                                \
+                jl_atomic_compare_exchange(&m ## _mutex, uv_thread_self(), 0); \
```

Comment:
I concur with improving the performance issues later.  If any of the locks turn out to be performance critical, we should probably first consider whether the higher level code can avoid it somehow.  (A fast lock is just a slightly shorter bottleneck. :-)